river/set.zig

```diff
@@ -0,0 +1,51 @@
+// This file is part of river, a dynamic tiling wayland compositor.
+//
+// Copyright 2022 The River Developers
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+const assert = @import("std").debug.assert;
+
+pub fn Set(comptime T: type, comptime capacity: usize) type {
+    return struct {
+        const Self = @This();
+
+        items: [capacity]T = undefined,
+        len: usize = 0,
+
+        pub fn add(self: *Self, new: T) void {
+            for (self.items[0..self.len]) |item| if (new == item) return;
+
+            assert(self.len < capacity);
```

Comment:
Is this ok to assert? Where is the current limit of 32 enforced?

Reply:
32 is a reasonable limit copied [from wlroots](https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/include/wlr/types/wlr_keyboard.h#L65) (i don’t know if there’s also a hard limit somewhere further down the stack).

There were three options:
- assume that it’s really never crossed and let runtime safety checks handle it
- assert and raise the limit once it’s crossed
- misbehave and don’t eat the key

Anyway, a comment would indeed do no harm.

Reply:
> assume that it’s really never crossed and let runtime safety checks handle it

Code written this way is by definition buggy IMO. Runtime safety checks (UB in release fast/small builds remember) should never be triggered by correct code.

> misbehave and don’t eat the key

This is what wlroots does, and I think it makes sense for river to do the same here since we depend on wlroots. Adding a comptime assertion that we match wlroot's limit wouldn't be a bad idea either:

```zig
comptime {
     assert(capacity == std.meta.fieldInfo(wlr.Keyboard, .keycodes).Array.len);
}
```

Reply:
An error log message if we do happen to misbehave due to this limit would make sense as well.

Reply:
Sorry, i can’t seem to make sense of this part of zig. It’s giving me `error: no member named 'Array' in struct 'std.builtin.StructField'`. How should it be?

Reply:
Yes that's right, sorry for the confusion.